doc/taler-exchange-dbinit.1
.Dd May 31, 2016
.Dt TALER-EXCHANGE-DBINIT 1
.Os
.Sh NAME
.Nm taler-exchange-dbinit
.Nd Initialize Taler exchange database.
.Sh SYNOPSIS
.Nm
.Op Fl d Ar DIRNAME | Fl -exchange-dir= Ns Ar DIRNAME
.Op Fl h | -help
.Op Fl g | -gc
.Op Fl r | -reset
.Op Fl v | -version
.Sh DESCRIPTION
.Nm
is a command line tool to initialize the Taler exchange database.
It creates the necessary tables and indices for the Taler exchange
to operate.
.Pp
Its options are as follows:
.Bl -tag -width indent
.It Fl d Ar DIRNAME | Fl -exchange-dir= Ns Ar DIRNAME
Use the configuration and other resources for the exchange to
operate from
.Ar DIRNAME .
.It Fl h | -help
Print short help on options.
.It Fl g | -gc
Garbage collect database.
Deletes all unnecessary data in the database.
.It Fl r | -reset
Drop tables.
Dangerous, will delete all existing data in the database before
creating the tables.
.It Fl v | -version
Print version information.
.El
.Sh SEE ALSO
.Xr taler-exchange-httpd 1 ,
.Xr taler-exchange-keyup 1 ,
.Xr taler-exchange-reservemod 1 ,
.Xr taler.conf 5 .
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt taler@gnu.org .


ChangeLog
Fri 03 May 2019 05:36:10 PM CEST
    Add support for compressing bodies during HTTP uploads to
    exchange and auditor. -CG

Thu 02 May 2019 09:15:37 PM CEST
    Pass hash of denomination public key in operations from
    clients to exchange instead of the (larger) full public key.
    Breaks protocol compatibility, bumping protocol to v3. -CG

Mon Mar 11 03:24:07 CET 2019
	Completed implementation of #5536 (delete private keys once we
	no longer need them). -CG

Sat Mar  2 19:09:43 CET 2019
	Changing denomination key revocation file format and moving them
	to their own directory (preparations for #5536 resolution). -CG

Sat Feb 23 14:47:12 CET 2019
	Actually disable 100 CONTINUE. -CG

Sun Aug 19 15:15:48 CEST 2018
	Increase various limits and rework transaction scopes to
	improve scalability. -CG

Tue Apr  3 23:29:06 CEST 2018
	Releasing Taler v0.5.0

Thu Jan  4 11:55:41 CET 2018
	Fix issue #5234 (aggregator ignoring refunds).
	Misc. minor fixes to the auditor. -CG

Mon Jan  1 23:15:37 CET 2018
	Add TALER_EXCHANGE_refund2() API call to libtalerexchange. -CG

Thu Dec 14 15:32:50 CET 2017
	Replaced taler-exchange-reservemod tool with new taler-bank-transfer
	tool (#5195).  Removed /admin/add/incoming API, replaced by new
	taler-exchange-wirewatch tooling. (#5077).

Sat Dec  9 15:21:50 CET 2071
	Implement refresh protocol optimization, bumping protocol version
	to v2 (#5178). -CG

Thu Nov  2 17:39:40 CET 2017
	Limit amount values to 2^53 as we always wanted (#5167). -CG

Thu Nov  2 16:50:46 CET 2017
	Fixing serious bug in reserve balance calculations, as
	fees were not always properly included in the summation. -CG

Tue Oct 31 13:43:55 CET 2017
	Extended crypto API to provide planchet generation functions
	explicitly (to be used on withdraw, refresh and tipping).
	Renamed a few API calls to better reflect their functions.
	Changed a few function signatures to use new structs. -CG

Wed Oct 18 15:20:29 CEST 2017
	Releasing Taler v0.4.0. -CG

Tue Oct 17 14:32:46 CEST 2017
	Fix building of /keys response to include full auditor
	signature set instead of partial map. -CG

Mon Oct 16 12:10:35 CEST 2017
	Realize JSON-based report formats from auditor (#4963). -CG

Sun Sep 17 16:46:13 CEST 2017
	Implement /keys cherry picking (#4840). -CG

Wed Sep 13 14:06:44 CEST 2017
	Add argument to TALER_EXCHANGE_check_keys_current()
	to force download even if /keys are still considered
	current. -CG

Fri Jul 14 17:38:54 CEST 2017
	Fix #4751 by not allowing denomination deletion (GC)
	to cascade into reserve_out table (and tolerating such
	deletion failures). This way, denominations will be kept
	around until the last reserve that references them (or
	known_coins derived from them) is closed. Naturally, in
	most case we expect reserves to be closed much faster
	than DKs, so in practice this should rarely apply. -CG

Sun Jun 11 17:03:56 CEST 2017
	Finish implementation and testing of automated refunding
	of expired reserves (#4956). -CG

Tue Jun  6 13:53:34 CEST 2017
	Releasing taler-exchange 0.3.0. -CG

Mon Apr 17 01:29:07 CEST 2017
	Add support for HTTP body compression (#4982). -CG

Mon Mar 20 04:37:46 CET 2017
	Implemented first working version of taler-auditor. -CG

Sat Mar 18 03:44:38 CET 2017
	Add 'wire_out' table to exchange DB to track outgoing
	wire transfers. -CG

Fri Nov 18 18:53:30 CET 2016
	Improved error reporting for bogus wire specifications.
	Releasing taler-exchange 0.2.0. -CG

Thu Nov 17 16:37:22 CET 2016
	Added missing idempotency checks for /refresh/reveal. -CG

Thu Nov 17 11:37:56 CET 2016
	Fixed a few cases of missing database rollbacks, causing the
	exchange to be stuck. -CG

Mon Jun 20 17:40:49 CEST 2016
 	Releasing taler-exchange 0.0.2. -CG

Mon Jun 13 16:35:09 CEST 2016
	Splitting HTTP service to run on different listen
	sockets for /admin and /test-APIs and the normal public
	APIs (#4569). -CG

Thu Jun  9 19:17:02 CEST 2016
	Rename /deposit/wtid to /track/transaction and
	/wire/deposits to /track/transfer. -CG

Wed Jun  8 11:32:41 CEST 2016
	Export libtalerfakebank API so we can use it in testcases for the
	merchant as well. -CG

Tue Jun  7 15:13:46 CEST 2016
        Adding public key of the exchange that was used to sign replies
        to applicable callbacks of libtalerexchange. (This will eventually
        be needed by the merchant's backend.) -CG

Mon Jun  6 23:59:05 CEST 2016
	Minor code cleanup, fixing FTBFS when run with just 'make'
	or when header installed to a location not included.
	Releasing taler-exchange 0.0.1. -CG

Wed Jun  1 17:27:36 CEST 2016
	Releasing taler-exchange 0.0.0. -CG


src/exchange-tools/taler-exchange-dbinit.c
/*
  This file is part of TALER
  Copyright (C) 2014, 2015 GNUnet e.V.

  TALER is free software; you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software
  Foundation; either version 3, or (at your option) any later version.

  TALER is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

  You should have received a copy of the GNU General Public License along with
  TALER; see the file COPYING.  If not, see <http://www.gnu.org/licenses/>
*/
/**
 * @file exchange-tools/taler-exchange-dbinit.c
 * @brief Create tables for the exchange database.
 * @author Florian Dold
 */
#include "platform.h"
#include <gnunet/gnunet_util_lib.h>
#include "taler_exchangedb_plugin.h"


/**
 * Return value from main().
 */
static int global_ret;

/**
 * -r option: do full DB reset
 */
static int reset_db;

/**
 * -g option: garbage collect DB reset
 */
static int gc_db;

/**
 * Main function that will be run.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  struct TALER_EXCHANGEDB_Plugin *plugin;

  if (NULL ==
      (plugin = TALER_EXCHANGEDB_plugin_load (cfg)))
  {
    fprintf (stderr,
             "Failed to initialize database plugin.\n");
    global_ret = 1;
    return;
  }
  if (reset_db)
    (void) plugin->drop_tables (plugin->cls);
  if (GNUNET_OK !=
      plugin->create_tables (plugin->cls))
  {
    fprintf (stderr,
             "Failed to initialize database.\n");
    TALER_EXCHANGEDB_plugin_unload (plugin);
    global_ret = 1;
    return;
  }
  if (gc_db)
    {
      if (GNUNET_SYSERR == plugin->gc (plugin->cls))
        fprintf (stderr,
                 "Garbage collection failed!\n");
    }
  TALER_EXCHANGEDB_plugin_unload (plugin);
}


/**
 * The main function of the database initialization tool.
 * Used to initialize the Taler Exchange's database.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc,
      char *const *argv)
{
  const struct GNUNET_GETOPT_CommandLineOption options[] = {
    GNUNET_GETOPT_option_flag ('r',
                                  "reset",
                                  "reset database (DANGEROUS: all existing data is lost!)",
                                  &reset_db),
    GNUNET_GETOPT_option_flag ('g',
                                  "gc",
                                  "garbage collect database",
                                  &gc_db),
    GNUNET_GETOPT_OPTION_END
  };

  /* force linker to link against libtalerutil; if we do
     not do this, the linker may "optimize" libtalerutil
     away and skip #TALER_OS_init(), which we do need */
  (void) TALER_project_data_default ();
  GNUNET_assert (GNUNET_OK ==
                 GNUNET_log_setup ("taler-exchange-dbinit",
                                   "INFO",
                                   NULL));
  if (GNUNET_OK !=
      GNUNET_PROGRAM_run (argc, argv,
                          "taler-exchange-dbinit",
			  "Initialize Taler exchange database",
			  options,
			  &run, NULL))
    return 1;
  return global_ret;
}


/* end of taler-exchange-dbinit.c */
